### PR TITLE
Spearbit 205

### PR DESCRIPTION
**Pull request by LayneHaber**
## Description

Fixes: https://github.com/spearbit-audits/connext-nxtp/issues/205

- doesn't use `_getCanonicalId` in `add` / `remove` router liquidity flows
- intentionally uses the `_local` as only the `_local` asset

## Type of change

- [ ] Docs change / dependency upgrade
- [ ] Configuration / tooling changes
- [ ] Refactoring
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Requires changes in customer code

## Related Issue(s)

Fixes

### CODE REVIEW

File: packages/deployments/contracts/contracts/core/connext/facets/RoutersFacet.sol [hunk 1]
```diff
@@ -564,7 +564,12 @@ contract RoutersFacet is BaseConnextFacet {
     if (_amount == 0) revert RoutersFacet__addLiquidityForRouter_amountIsZero();
 
     // Get the canonical asset ID from the representation.
-    (TokenId memory canonical, bytes32 key) = _getApprovedCanonicalId(_local);
+    // NOTE: not using `_getApprovedCanonicalId` because candidate can *only* be local
+    TokenId memory canonical = s.representationToCanonical[_local];
```
Comment (Saw-mon-and-Natalie) [inline packages/deployments/contracts/contracts/core/connext/facets/RoutersFacet.sol:568]: `_local` can be a canonical asset on its canonical domain which then `s.representationToCanonical[_local]` will not be set (is the default value). In this case the values for `canonical` and `key` will not be correctly calculated.

File: packages/deployments/contracts/contracts/core/connext/facets/RoutersFacet.sol [hunk 2]
```diff
@@ -615,7 +620,8 @@ contract RoutersFacet is BaseConnextFacet {
 
     // Get the canonical asset ID from the representation.
     // NOTE: allow getting unapproved assets to prevent lockup on approval status change
-    TokenId memory canonical = _getCanonicalTokenId(_local);
+    // NOTE: not using `_getCanonicalTokenId` because candidate can *only* be local
+    TokenId memory canonical = s.representationToCanonical[_local];
```
Comment (Saw-mon-and-Natalie) [inline packages/deployments/contracts/contracts/core/connext/facets/RoutersFacet.sol:624]: See issue as https://github.com/connext/nxtp/pull/2489#discussion_r1025708218

File: packages/deployments/contracts/contracts/core/connext/facets/RoutersFacet.sol [hunk 3]
```diff
@@ -563,7 +568,18 @@ contract RoutersFacet is BaseConnextFacet {
...
+    // NOTE: not using `_getApprovedCanonicalId` because candidate can *only* be local
+    TokenId memory canonical = s.representationToCanonical[_local];
+    if (canonical.domain == 0) {
+      // Assume you are on the canonical domain, which does not update the above mapping
+      // If this is an incorrect assumption, the approval should fail
+      canonical.domain = s.domain;
+      canonical.id = TypeCasts.addressToBytes32(_local);
+    }
+    bytes32 key = AssetLogic.calculateCanonicalHash(canonical.id, canonical.domain);
+    if (!_isAssetWhitelistRemoved() && !s.approvedAssets[key]) {
```
Comment (Saw-mon-and-Natalie) [inline packages/deployments/contracts/contracts/core/connext/facets/RoutersFacet.sol:580]: If `_isAssetWhitelistRemoved() == true`, then for assets that have not been added to the protocol we are assuming:

1. The canonical domain is the current domain
2. The canonical id is basically the `_local` provided

And then using these incorrect values, we calculate `key` and allow routers to deposit liquidity and finally emit `RouterLiquidityAdded` using the incorrect `key`.

File: packages/deployments/contracts/contracts/core/connext/facets/RoutersFacet.sol [hunk 4]
```diff
@@ -562,7 +567,18 @@ contract RoutersFacet is BaseConnextFacet {
     if (_amount == 0) revert RoutersFacet__addLiquidityForRouter_amountIsZero();
 
     // Get the canonical asset ID from the representation.
-    (TokenId memory canonical, bytes32 key) = _getApprovedCanonicalId(_local);
+    // NOTE: not using `_getApprovedCanonicalId` because candidate can *only* be local
+    TokenId memory canonical = s.representationToCanonical[_local];
+    if (canonical.domain == 0) {
```
Comment (Saw-mon-and-Natalie) [inline packages/deployments/contracts/contracts/core/connext/facets/RoutersFacet.sol:572]: Here we are assuming that when `canonical.domain == 0`, `canonical.id` is also not set. Perhaps we should check this.

File: packages/deployments/contracts/contracts/core/connext/libraries/LibConnextStorage.sol [hunk 5]
```diff
@@ -238,10 +238,6 @@ struct AppStorage {
   bool _routerAllowlistRemoved;
   // 25
   uint256 _routerAllowlistTimestamp;
-  // 26
-  bool _assetAllowlistRemoved;
-  // 27
-  uint256 _assetAllowlistTimestamp;
```
[the remark this answers is not in the document]
Comment (LayneHaber) [reply packages/deployments/contracts/contracts/core/connext/libraries/LibConnextStorage.sol:244]: yes, acknowledged!

[the remark this answers is not in the document]
Comment (LayneHaber) [reply packages/deployments/contracts/contracts/core/connext/libraries/LibConnextStorage.sol:244]: added a comment: https://github.com/connext/nxtp/pull/2489/commits/ff3e81cc110e5662a5da5060421fbb17f963b0c3

File: packages/deployments/contracts/contracts/core/connext/facets/RoutersFacet.sol (hunk 4 above)
Comment (LayneHaber) [reply packages/deployments/contracts/contracts/core/connext/facets/RoutersFacet.sol]: added! https://github.com/connext/nxtp/pull/2489/commits/1afe00a3e5bddbf0424cf954b1f77a807b1aa6d5

**Pull request merged**